File: library-java/src/main/java/com/github/shiraji/butai/ButaiActivityLifecycleCallbacks.java
package com.github.shiraji.butai;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

public class ButaiActivityLifecycleCallbacks implements Application.ActivityLifecycleCallbacks, Butai {

    private int runningActivities = 0;

    private AppStatus appStatus = AppStatus.JUST_APPEARED;

    @Override
    public void onActivityStarted(Activity activity) {
        if (runningActivities == 0) appStatus = AppStatus.JUST_APPEARED;
        else if (runningActivities > 0) appStatus = AppStatus.FOREGROUND;
        else appStatus = AppStatus.BACKGROUND;
        runningActivities++;
    }

    @Override
    public void onActivityStopped(Activity activity) {
        runningActivities--;
        if (runningActivities <= 0) {
            appStatus = AppStatus.BACKGROUND;
            runningActivities = 0;
        }
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle bundle) {
    }

    @Override
    public void onActivityResumed(Activity activity) {
    }

    @Override
    public void onActivityPaused(Activity activity) {
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle bundle) {
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
    }

    @Override
    public boolean isReturnedFromBackground() {
        return appStatus == AppStatus.JUST_APPEARED;
    }

    @Override
    public boolean isBackground() {
        return !isForeground();
    }

    @Override
    public boolean isForeground() {
        return appStatus.isForeground;
    }
}

File: library/src/main/java/com/github/shiraji/butai/ButaiActivityLifecycleCallbacks.kt
package com.github.shiraji.butai

import android.app.Activity
import android.app.Application
import android.os.Bundle

internal class ButaiActivityLifecycleCallbacks : Application.ActivityLifecycleCallbacks, Butai {

    private val delegate = ButaiActivityLifecycleCallbacksDelegate()

    override fun onActivityStarted(p0: Activity?) = delegate.onActivityStarted()
    override fun onActivityStopped(p0: Activity?) = delegate.onActivityStopped()
    override fun onActivityPaused(p0: Activity?) = Unit
    override fun onActivityDestroyed(p0: Activity?) = Unit
    override fun onActivitySaveInstanceState(p0: Activity?, p1: Bundle?) = Unit
    override fun onActivityCreated(p0: Activity?, p1: Bundle?) = Unit
    override fun onActivityResumed(p0: Activity?) = Unit
    override fun isReturnedFromBackground() = delegate.isReturnedFromBackground
    override fun isBackground() = delegate.isBackground
    override fun isForeground() = delegate.isForeground
}